x-pack/plugin/data-streams/src/main/java/org/elasticsearch/xpack/datastreams/UpdateTimeSeriesRangeService.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.datastreams;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateTaskExecutor;
+import org.elasticsearch.cluster.ClusterStateUpdateTask;
+import org.elasticsearch.cluster.LocalNodeMasterListener;
+import org.elasticsearch.cluster.metadata.DataStream;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Priority;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+
+import static org.elasticsearch.index.mapper.DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER;
+
+/**
+ * A component that updates the 'index.time_series.end_time' index setting for the most recently added backing index of a tsdb data stream.
+ */
+public class UpdateTimeSeriesRangeService extends AbstractLifecycleComponent implements LocalNodeMasterListener {
+
+    private static final Logger LOGGER = LogManager.getLogger(UpdateTimeSeriesRangeService.class);
+
+    private final TimeValue pollInterval;
+    private final ThreadPool threadPool;
+    private final ClusterService clusterService;
+
+    private volatile Scheduler.Cancellable job;
+
+    UpdateTimeSeriesRangeService(Settings settings, ThreadPool threadPool, ClusterService clusterService) {
+        this.pollInterval = DataStreamsPlugin.TIME_SERIES_POLL_INTERVAL.get(settings);
+        this.threadPool = threadPool;
+        this.clusterService = clusterService;
+    }
+
+    void perform() {
+        job = null;
+        clusterService.submitStateUpdateTask("update_tsdb_data_stream_end_times", new ClusterStateUpdateTask(Priority.URGENT) {
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                return updateTimeSeriesTemporalRange(currentState, Instant.now());
+            }
+
+            @Override
+            public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                scheduleTask();
```

Comment:
Why did you decide to go with a manual re-scheduling for this task instead of using something like the `SchedulerEngine` that only has to be scheduled once?

Reply:
To avoid a situation where one or more cluster state update tasks that update data streams are scheduled while an existing one is still scheduled or executing. From that perspective, it made sense to me, to schedule an update once and only reschedule another one when current update has completed.